File: src/main/java/com/team04/back/domain/cloth/cloth/controller/ClothController.java
package com.team04.back.domain.cloth.cloth.controller;

import com.team04.back.domain.cloth.cloth.dto.CategoryClothDto;
import com.team04.back.domain.cloth.cloth.dto.ExtraClothDto;
import com.team04.back.domain.cloth.cloth.dto.OutfitResponse;
import com.team04.back.domain.cloth.cloth.dto.WeatherClothResponseDto;
import com.team04.back.domain.cloth.cloth.entity.Clothing;
import com.team04.back.domain.cloth.cloth.entity.ExtraCloth;
import com.team04.back.domain.cloth.cloth.enums.Category;
import com.team04.back.domain.cloth.cloth.service.ClothService;
import com.team04.back.domain.weather.weather.dto.WeatherInfoDto;
import com.team04.back.domain.weather.weather.entity.WeatherInfo;
import com.team04.back.domain.weather.weather.service.WeatherService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import jakarta.validation.constraints.FutureOrPresent;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/v1/cloth")
@RequiredArgsConstructor
public class ClothController {
    private final ClothService clothService;
    private final WeatherService weatherService;

    @GetMapping("/details")
    @Operation(summary = "날씨 기반 옷 정보 조회", description = "위도와 경도를 이용하여 날씨 정보를 조회하고, 해당 날씨에 적합한 옷 정보를 반환합니다.")
    public WeatherClothResponseDto getClothDetails(
            @Parameter(description = "위도", example = "37.5")
            @RequestParam(name = "latitude") double latitude,

            @Parameter(description = "경도", example = "127.0")
            @RequestParam(name = "longitude") double longitude) {
        //좌표 기반으로 날씨 정보 가져오기
        WeatherInfo weatherInfo = weatherService.getWeatherInfo(latitude, longitude, LocalDate.now());

        // 날씨 정보에 따라 옷 정보 가져오기
        List<CategoryClothDto> cloths = clothService.findClothByWeather(weatherInfo.getFeelsLikeTemperature());

        // 날씨 정보에 따라 추가 옷 정보 가져오기
        Set<ExtraCloth> extraCloth = clothService.getExtraClothes(weatherInfo);

        // 날씨 정보와 옷 정보를 포함한 응답 DTO 리턴
        WeatherInfoDto weatherInfoDto = new WeatherInfoDto(weatherInfo);
        Set<ExtraClothDto> extraClothDto  =extraCloth.stream()
                .map(ExtraClothDto::new)
                .collect(Collectors.toSet());
        return new WeatherClothResponseDto(weatherInfoDto, cloths, extraClothDto);
    }

    record TripSchedule(
            @FutureOrPresent
            LocalDate start,
            @FutureOrPresent
            LocalDate end,
            @Min(-90) @Max(90)
            double lat,
            @Min(-180) @Max(180)
            double lon
    ) {
        public TripSchedule {
            if (start != null && end != null) {
                if (!start.isBefore(end)) {
                    throw new IllegalArgumentException("시작 날짜는 종료 날짜보다 이전이어야 합니다.");
                }
                if (ChronoUnit.DAYS.between(start, end) > 30) {
                    throw new IllegalArgumentException("최대 30일까지 조회할 수 있습니다.");
                }
            }
        }
    }

    @GetMapping
    public OutfitResponse getOutfitWithPeriod(TripSchedule tripSchedule) {
        List<WeatherInfo> duration = weatherService.getWeatherInfos(
                tripSchedule.lat,
                tripSchedule.lon,
                tripSchedule.start,
                tripSchedule.end
        );
        Map<Category, List<Clothing>> outfits = clothService.getOutfitWithPeriod(duration);
        return new OutfitResponse(outfits);
    }
}

File: src/main/java/com/team04/back/domain/cloth/cloth/controller/ClothController.kt
package com.team04.back.domain.cloth.cloth.controller

import com.team04.back.domain.cloth.cloth.dto.OutfitRecommendationResponseDto
import com.team04.back.domain.cloth.cloth.dto.WeatherClothResponseDto
import com.team04.back.domain.cloth.cloth.service.ClothService
import com.team04.back.domain.cloth.cloth.service.TopOnlyClothService
import com.team04.back.domain.weather.geo.service.GeoService
import com.team04.back.domain.weather.weather.service.WeatherService
import io.swagger.v3.oas.annotations.Operation
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.time.LocalDate

@RestController
@RequestMapping("/api/v1/cloth")
class ClothController(
    private val clothService: ClothService,
    private val weatherService: WeatherService,
    private val geoService: GeoService,
    private val topOnlyClothService: TopOnlyClothService,

    ) {
    @GetMapping("/details")
    @Operation(summary = "날씨 기반 옷 정보 조회", description = "위도와 경도를 이용하여 날씨 정보를 조회하고, 해당 날씨에 적합한 옷 정보를 반환합니다.")
    fun getClothDetails(
        @RequestParam latitude: Double,
        @RequestParam longitude: Double,
    ): WeatherClothResponseDto  {
        // 1. 위도, 경도로 위치 조회
        val location = geoService.getLocationFromCoordinates(latitude, longitude)

        // 2. 현재 날씨 조회
        val weatherInfo = weatherService.getWeatherInfo(latitude, longitude, LocalDate.now())

        // 3. 오늘 날짜 범위로 날씨 계획을 만들어서 서비스 호출
        val weatherPlan = weatherService.getWeatherInfos(latitude, longitude, LocalDate.now(), LocalDate.now())


        // 4. 현재 날씨에 맞는 옷차림 추천 조회
        val outfitRecommendations : OutfitRecommendationResponseDto = topOnlyClothService.getOutfitRecommendations(weatherPlan, location)

        return WeatherClothResponseDto(
            weatherInfo = weatherInfo,
            recommendedOutfits = outfitRecommendations.recommendedOutfits,
            notRecommendedOutfits = outfitRecommendations.notRecommendedOutfits
        )
    }

    @GetMapping
    @Operation(summary = "기간별 옷차림 조회", description = "시작일과 종료일을 이용하여 해당 기간 동안의 날씨에 적합한 옷차림 정보를 반환합니다.")
    fun getOutfitWithPeriod(
        @RequestParam latitude: Double,
        @RequestParam longitude: Double,
        @RequestParam location: String,
        @RequestParam startDate: LocalDate,
        @RequestParam endDate: LocalDate
    ): OutfitRecommendationResponseDto {
        // 1. 기간 동안 날씨 계획 조회
        val weatherPlan = weatherService.getWeatherInfos(
            latitude,
            longitude,
            startDate,
            endDate
        )

        // 2. ClothService에서 추천/비추천 의류 계산
        val outfitRecommendations : OutfitRecommendationResponseDto = clothService.getOutfitRecommendations(weatherPlan, location)
        return outfitRecommendations
    }
}
